### ISSUE TITLE
Call function on every new adbapi connection

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10364]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#480|
|Type|enhancement|
|Created|2004-01-17 03:29:31Z|

Attachments:

* [enterprise.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/615/6153f0b97fb44420cd413a37979f8e6219fc3f36/767261c546360c1ef5ded52bb2b6e04d37a5b7a7.diff) (7017 bytes) - added by jknight on 2004-05-29 09:32:44Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__480 480
type__enhancement enhancement
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1074310171000000 1074310171000000
changetime__1470879118838325 1470879118838325
version__ 
owner__davep davep
cc__davep cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
This patch adds an optional parameter  to a dbpool that gets called whenever a
new connection is opened. This is useful to setup per-connection parameters,
such as charset and timezone.

I use it as follows:
====
def db_set_encoding(con):
    cursor = con.cursor()
    cursor.execute("SET CLIENT_ENCODING TO UNICODE")
    
site.dbpool = adbapi.ConnectionPool("pyPgSQL.PgSQL",
database="db",client_encoding=("utf-8", "ignore"), unicode_results=1,
cp_openfun=db_set_encoding)
====

Index: twisted/enterprise/adbapi.py
===================================================================
RCS file: /cvs/Twisted/twisted/enterprise/adbapi.py,v
retrieving revision 1.58
diff -u -r1.58 adbapi.py
--- twisted/enterprise/adbapi.py        24 Nov 2003 00:55:28 -0000      1.58
+++ twisted/enterprise/adbapi.py        16 Jan 2004 21:28:15 -0000
@@ -56,6 +56,10 @@
     number of connections that will be opened by the pool. You can pass
     the noisy arg which determines whether informational log messages are
     generated during the pool's operation.
+
+    You can pass a function cp_openfun which will get called after every
+    .connect() operation on the underlying db api. This is useful to setup
+    per-connection state, such as charset, timezone, etc.
     """
 
     noisy = 1   # if true, generate informational log messages
@@ -63,6 +67,8 @@
     max = 5     # maximum number of connections in pool
     running = 0 # true when the pool is operating
 
+    openfun = None # A function to call to setup a new connection
+    
     def __init__(self, dbapiName, *connargs, **connkw):
         self.dbapiName = dbapiName
         self.dbapi = reflect.namedModule(dbapiName)
@@ -88,6 +94,10 @@
             self.noisy = connkw['cp_noisy']
             del connkw['cp_noisy']
 
+        if connkw.has_key('cp_openfun'):
+            self.openfun = connkw['cp_openfun']
+            del connkw['cp_openfun']
+        
         self.min = min(self.min, self.max)
         self.max = max(self.min, self.max)
 
@@ -224,6 +234,8 @@
                                                         self.connargs or '',
                                                         self.connkw or ''))
             conn = apply(self.dbapi.connect, self.connargs, self.connkw)
+            if self.openfun != None:
+                self.openfun(conn)
             self.connections[tid] = conn
         return conn
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Attached new patch with tests. Davep: can you please check&amp;apply?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
I don't understand why disconnect() was added. What is the use case?
Is it just for the test? Also, the test for the connect callback seems
to depend on the order that tests are run, yes? Not so nice.

Finally this seems to break the gadfly test, though I haven't
determined why.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Oops I should have explained that.
The use case for disconnect is when you want to do some synchronous database work
either in your own thread (in the midst of some other stuff you're doing), or else in the
main thread, while the app is still starting up and thus blocking doesn't matter.

So, you can call dbpool.connect() to get a dbhandle for your thread (with the same user/pass/
arguments/etc as the dbpool would get) and then do synchronous operations in the midst of
whatever you're doing. When you're done, you need to be able to close down the database 
connection, thus disconnect().

For gadfly: perhaps it doesn't support >1 simultaneous connection to the database? In which
case it's the testcase that's broken, for it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
Ok, I see.

I would like to see the test for the open callback to be
independent of the order that tests are run. Also, right
now it just seems to test whether it is called at all,
and not per-connection, but maybe that's not a big deal.

So, could you leave the existing testPool alone and do
the connect/disconnect tests in testOpenfun, fixing the
gadfly test as needed?

I have some reservations about disconnect, because I would
like to move away from associating connections with threads.
Would you consider making disconnect take the connection as
an argument?

Also, please make the documentation for disconnect clear
on the fact that it is only to be used in tandem with
connect() during synchronous operation, and not during
normal asynchronous usage.

thanks,
dave
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
I guess your testOpenfun depends on startUp making a connection,
not the order of the tests? In either case, I still would like
to see the new connect/disconnect + sync operation test in the
new test method.

For gadfly, maybe you can skip the connect(), disconnect(),
or maybe close the existing connection first?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
I have some time this weekend, I'll work up something on this.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)|@jdavisp3 commented|
|-|-|

```
#!html
<pre>
This is in svn now.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|GitHub commented|
|-|-|

In changeset f42ff53e4057460b0337333fb04200b1a4d8b86e
```
#!CommitTicketReference repository="" revision="f42ff53e4057460b0337333fb04200b1a4d8b86e"
Merge pull request #480 from twisted/8738-rodrigc-words-py3

Author: rodrigc
Reviewer: alex
Fixes: #8738

Add missing twisted.words entries to dist3.py so that Python 3 tests run
```
